feat(web): Remove from Stack

xCJPECKOVERx opened the pull request:
## Description

Adds a menu option for stack assets to remove them from the stack. Only available on stacks with more than 2 assets, and not on the primary asset.
Works by calling stack.create, and adds a 'merge' field to the StackCreateDto. When removing from the stack is does not merge in the assets of existing stack children. Ensures that when adding to or creating a stack, merge is true.
When the asset is removed, the asset-viewer's asset is set to the stack's primary asset.

Updated the icon for Unstack, so that remove can use the image-minus icon.

## How Has This Been Tested?

Create a stack, choose a stack asset that's not the primary, click 'remove from stack' in the navbar menu. The asset-viewer stack will be updated, and the asset-viewer will display the stack's primary asset. Closing the viewer will show the stack count updated, and the removed asset visible in the timeline.

<details><summary><h2>Screenshots (if appropriate)</h2></summary>

<!-- Images go below this line. -->
<img width="222" alt="image" src="https://github.com/user-attachments/assets/d2e09f42-f8f1-4059-9e46-8f01691d1242" />


</details>

## Checklist:

- [x] I have performed a self-review of my own code
- [x] I have made corresponding changes to the documentation if applicable
- [x] I have no unrelated changes in the PR.
- [x] I have confirmed that any new dependencies are strictly necessary.
- [x] I have written tests for new code (if applicable)
- [x] I have followed naming conventions/patterns in the surrounding code
- [x] All code in `src/services/` uses repositories implementations for database calls, filesystem operations, etc.
- [x] All code in `src/repositories/` is pretty basic/simple and does not have any immich specific logic (that belongs in `src/services/`)

xCJPECKOVERx committed:
- add component
- update server's StackCreateDto for merge parameter
- Update stackRepo to only merge stacks when merge=true (default)
- update web action handlers to show stack changes

xCJPECKOVERx committed:
- make open-api

xCJPECKOVERx committed:
lint & format

xCJPECKOVERx committed:
- Add proper icon to 'remove from stack'
- change web unstack icon to image-off-outline

xCJPECKOVERx committed:
Merge branch 'main' into feat/remove-from-stack

xCJPECKOVERx committed:
- cleanup

xCJPECKOVERx committed:
- format & lint

xCJPECKOVERx committed:
- make open-api: StackCreateDto merge optional

danieldietzler reviewed (commented):
I feel like an endpoint like `DELETE stacks/:id/asset/:id` would be better? @jrasm91 thoughts?

xCJPECKOVERx commented:
Maybe that or the asset's update endpoint to clear the stack id? I used create since that's how updates to the stack (adding to it) are handled already: delete the existing stacks and create a new one with the required assets.

danieldietzler commented:
Yeah I saw that. I just don't really like the ergonomics of having to create a new stack. That's similar to extending the update endpoint, there we'd also need to enumerate all stack assets again which seems kind of annoying to deal with. That's why I was suggesting such a DELETE endpoint

xCJPECKOVERx commented:
Sounds good, I can work on it over the next week or so. 
For the second-to-last asset issue, I agree with an error. On front end my plan is to not allow removal if there's only 2 assets anyway.

Another edge case is removing the primary asset, if allowed. I'm currently not allowing it, though arguably wouldn't be too hard to handle in a new endpoint if it can optionally return the new primary asset for the front end to display.

danieldietzler commented:
You could even just have the frontend re-fetch the stack anyways after deleting that asset

xCJPECKOVERx commented:
That would depend on whether we want to keep viewing the stack (update the viewing asset), or stay on the old primary which is no longer a part of the stack.

xCJPECKOVERx committed:
initial addition of new endpoint

xCJPECKOVERx committed:
Merge branch 'main' into feat/remove-from-stack

xCJPECKOVERx committed:
Merge branch 'main' into feat/remove-from-stack

xCJPECKOVERx committed:
remove stack endpoint

xCJPECKOVERx committed:
- fix up remove stack endpoint
- open-api

xCJPECKOVERx committed:
Merge branch 'main' into feat/remove-from-stack

xCJPECKOVERx committed:
- Undo stackCreate merge parameter

xCJPECKOVERx committed:
- open-api typescript

xCJPECKOVERx committed:
open-api dart

xCJPECKOVERx committed:
Tests:
- add tests
- update assetStub.imageFrom2015 to have required stack attributes to include it with tests

xCJPECKOVERx committed:
Merge branch 'main' into feat/remove-from-stack

xCJPECKOVERx committed:
update event name

xCJPECKOVERx committed:
Fix event name in test

xCJPECKOVERx commented on web/src/lib/components/photos-page/asset-grid.svelte:
```diff
@@ -523,6 +523,23 @@
+      case AssetAction.REMOVE_ASSET_FROM_STACK: {
+        timelineManager.addAssets([toTimelineAsset(action.asset)]);
+        if (action.stack) {
+          //Have to unstack then restack assets in timeline in order to update the stack count in the timeline.
+          updateUnstackedAssetInTimeline(
+            timelineManager,
+            action.stack.assets.map((asset) => toTimelineAsset(asset)),
+          );
+          updateStackedAssetInTimeline(timelineManager, {
+            stack: action.stack,
+            toDeleteIds: action.stack.assets
+              .filter((asset) => asset.id !== action.stack?.primaryAssetId)
+              .map((asset) => asset.id),
+          });
```
I could take a look at this in a separate PR. No super knowledgeable in how the timeline manager works quite yet.

xCJPECKOVERx committed:
remove asset_update check

xCJPECKOVERx commented on server/src/services/stack.service.ts:
```diff
@@ -58,6 +58,40 @@ export class StackService extends BaseService {
+    await this.requireAccess({ auth, permission: Permission.ASSET_UPDATE, ids: [assetId] });
+
+    //Verify the asset is in the stack
+    const asset = await this.assetRepository.getById(assetId);
+    if (!asset) {
+      throw new BadRequestException('Asset not found');
+    }
+    if (asset?.stackId !== stackId) {
+      throw new BadRequestException('Asset not in stack');
+    }
+
+    //Verify the stack has more than 2 assets
+    const stack = await this.findOrFail(stackId);
+    if (stack.assets?.length <= 2) {
+      throw new BadRequestException('Cannot remove last or second to last asset');
+    }
+
+    //Verify the asset is not the stack's primary asset
+    if (stack.primaryAssetId === assetId) {
+      throw new BadRequestException("Cannot remove stack's primary asset");
+    }
+
+    await this.assetRepository.update({ id: assetId, stackId: null });
```
> combine some of the selects

Just to confirm correct procedure: for helper queries like this, I should just add a repository function to return a custom (or existing dto) object? 
For this particular case though, I could remove the asset fetch and just get the stack and verify the asset's in there. If the asset doesn't exist, it wouldn't be in the stack anyway.

> I'm not sure if we need to do the <= 2 assets check

If we don't require it, then there should be some stack cleanup. imo it's yucky having an asset showing as a stack of 1. Alternatively, we could remove the check here, but still hide the menu option if there's only 2 in the stack.

jrasm91 replied:
For the helper query, just add a new repository method on the `StackRepository` and you don't need an explicit return type. You can see how the other methods work, especially the asset job repository. You can dynamically select the data you need. I'm also happy to do this for you if you don't want to bother with it.

My point was mainly that we don't need a server-side check here since there are still other ways the user can get into that state. I agree it would be good to add a nightly job or similar that takes care of cleaning up this type of data.

xCJPECKOVERx replied:
Gotcha. Currently reworking it a bit and I can omit the asset fetch entirely. With that being the case, is a new function really necessary/performative enough over just getting the stack with the assets (like in update's primaryAssetId check)?

jrasm91 commented on server/src/services/stack.service.ts:
```diff
@@ -58,6 +58,40 @@ export class StackService extends BaseService {
+    await this.assetRepository.update({ id: assetId, stackId: null });
+    await this.eventRepository.emit('StackUpdate', { stackId, userId: auth.user.id });
+
+    const updatedStack = {
+      ...stack,
+      assets: stack.assets?.filter((a) => a.id !== assetId),
+    };
+    return mapStack(updatedStack, { auth });
```
Yeah, it should have enough information client-side to figure out what needs to be updated.

xCJPECKOVERx committed:
- merge stack.removeAsset params into one object
- refactor asset existence check (no need for asset fetch)
- fix tests

xCJPECKOVERx committed:
Don't return updated stack

xCJPECKOVERx committed:
Create specialized stack id & primary asset fetch for asset removal checks

xCJPECKOVERx committed:
Merge branch 'main' into feat/remove-from-stack

xCJPECKOVERx committed:
Correct new permission names

xCJPECKOVERx committed:
make sql

xCJPECKOVERx committed:
Merge branch 'main' into feat/remove-from-stack

xCJPECKOVERx committed:
- fix open-api

jrasm91 commented on server/src/services/stack.service.ts:
```diff
@@ -58,6 +59,26 @@ export class StackService extends BaseService {
+
+    const stack = await this.stackRepository.getForAssetRemoval(assetId);
+
+    //Verify the asset exists and is in the stack
```
We can remove these two comments

xCJPECKOVERx committed:
- cleanup

xCJPECKOVERx committed:
Merge branch 'main' into feat/remove-from-stack
